update InvalidPackageDeclaration to report if rootPackage is not present

mateot1 opened the pull request:
This should fix the issue I just opened here: https://github.com/detekt/detekt/issues/4483

marschwar commented on detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/InvalidPackageDeclaration.kt:
```diff
@@ -34,26 +34,28 @@ class InvalidPackageDeclaration(config: Config = Config.empty) : Rule(config) {
-                    declaredPath
-                } else {
-                    declaredPath.substringAfter(normalizedRootPackage)
+            val expectedPath = when {
+                normalizedRootPackage.isBlank() -> declaredPath
+                !declaredPath.startsWith(normalizedRootPackage) -> {
+                    directive.reportInvalidPackageDeclaration("The package declaration is missing the root package")
+                    return
                 }
+                else -> declaredPath.substringAfter(normalizedRootPackage)
+            }
```
If find a return statement within a when a little bit confusing. I would prefer the more verbose but also more explicit version:

```kotlin
if (normalizedRootPackage.isNotBlank() && !declaredPath.startsWith(normalizedRootPackage)) {
    directive.reportInvalidPackageDeclaration("The package declaration is missing the root package")
    return
}
val expectedPath = if (normalizedRootPackage.isBlank()) {
    declaredPath
} else {
    declaredPath.substringAfter(normalizedRootPackage)
}
```

cortinico commented on detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/InvalidPackageDeclaration.kt:
```diff
@@ -25,7 +25,7 @@ class InvalidPackageDeclaration(config: Config = Config.empty) : Rule(config) {
     )
 
-    @Configuration("if specified this part of the package structure is ignored")
+    @Configuration("if specified the declaration must start with this root, but it is ignored in the file structure")
```
Why "it is ignored in the file structure"?

mateot1 replied:
I guess I was just trying to reword the existing behavior and used "file" instead of "package"... this change forces you to include the root in the package _declaration_ but you can still have a file structure that doesn't include those root directories. e.g. `package my.root.package.file.path` works with `file/path` as well as `my/root/package/file/path`

marschwar replied:
I guess that is the problem when a single configuration means multiple things. ;) How about:

_If specified the package declaration must start with this base package. In the directory structure this part may or may not be omitted._

Also it may help if we include a link to the [corresponding section in the Kotlin Coding Conventions](https://kotlinlang.org/docs/coding-conventions.html#directory-structure).

BraisGabin replied:
I think that we shouldn't "overload" this configuration. If we want that all the files start with the same package we should create a new configuration for it. And, if you want to ignore the first part of that root path AND enforce it you need to set it twice.

mateot1 replied:
okay, so just like a `requireRootInDeclaration` boolean?

BraisGabin replied:
Sorry for the late answer. Yes, I think that a `requireRootInDeclaration` have sense.

mateot1 commented:
sorry for the delay. I updated this to use a new config field and added a default to `default-detekt-config.yml`. let me know if there's anything I missed for adding a new config field.

cortinico commented on detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/InvalidPackageDeclarationSpec.kt:
```diff
@@ -106,6 +108,37 @@ internal class InvalidPackageDeclarationSpec : Spek({
+                    class C
+                """
+
+                val ktFile = compileContentForTest(source, createPath("src/foo/bar/File.kt"))
```
Can you add another test for the file in the following path:
```
src/com/example/foo/bar/File.kt
```
